File: vertor/ArrayVector.go
package vertor

import (
	"dataStruct"
	"errors"
)

const (
	defaultCapacity = 10
)

type ArrayVector struct {

	// 当前向量的长度
	capacity int

	// 存放数据的容器
	container [...]dataStruct.Person

}

// 创建ArrayVector实例对象
func NewArrayVector(capacity int) *ArrayVector {
	arrayVector := ArrayVector{}
	arrayVector.capacity = capacity
	arrayVector.container = [capacity]dataStruct.Person{}
	return &arrayVector
}

// 默认长度的ArrayVector实例对象
func NewArrayVectorDefault() *ArrayVector {
	return NewArrayVector(defaultCapacity)
}

// 增长容器的空间
func (av *ArrayVector) grow() {
	newVector := NewArrayVector(av.capacity * 0.5)
	newVector.copy(av)
	av = newVector
}

// 复制元素
func (av *ArrayVector) copy(old *ArrayVector) {
	for i := 0; i < old.capacity; i++ {
		av.container[i] = old.container[i]
	}
}

// 获取向量规模大小
func (av *ArrayVector) Size() int {
	return av.capacity
}

// 获取指定位置的元素
func (av *ArrayVector) Get(index int) (dataStruct.Person, error){
	if index >= 0 && index < av.capacity {
		return av.container[index], nil
	}
	return dataStruct.Person{}, errors.New("not found")
}

// 替换指定位置上的元素
func (av *ArrayVector) Put(index int, person dataStruct.Person) error {
	if index >= av.capacity {
		return errors.New("out of index")
	}
	av.container[index] = person
	return nil
}

File: src/vector/ArrayVector.java
package vector;

/**
 * 连续结构的Vector
 * Created by eric on 17-10-31
 */
public class ArrayVector implements Vector {

    private static final int defaultSize = 0;
    private static final int defaultCapacity = 10;

    // 记录已使用的元素数量
    private int size;

    // 容器的最大容量
    private int capacity;

    // 容器
    private Person[] container;

    public ArrayVector() {
        this.size = defaultSize;
        this.capacity = defaultCapacity;
        this.container = new Person[defaultCapacity];
    }


    @Override
    public int size() {
        return container.length;
    }

    @Override
    public Person get(int index) {
        if (index >= capacity) {
            throw new IndexOutOfBoundsException("数组越界");
        }
        return container[index];
    }

    @Override
    public void put(int index, Person person) {

    }

    @Override
    public void insert(int index, Person person) {

    }

    @Override
    public void remove(int index) {

    }

    @Override
    public void remove(int start, int end) {

    }

    @Override
    public void sort() {

    }

    @Override
    public int find(Person person) {
        return 0;
    }

    @Override
    public void deduplicate() {

    }

    /**
     * 检查是否需要增长长度
     * @param index
     */
    private void checkCapacity(int index) {
        if (index >= capacity || size >= capacity) {
            grow();
        }
    }

    /**
     * 自增长容器长度
     */
    private void grow() {
        int newCapacity = capacity * 2;
        Person[] newContainer = new Person[newCapacity];
        for (int i = 0; i < capacity; i++) {
            newContainer[i] = container[i];
        }
        capacity = newCapacity;
        container = newContainer;
    }
}
